appinventor/blocklyeditor/src/blocks/text.js

```diff
@@ -317,21 +317,126 @@ Blockly.Blocks['text_starts_at'] = {
+        Blockly.Msg.LANG_TEXT_CONTAINS_INPUT_PIECE);
+
+    this.setOutput(true, getType("boolean", utils.OUTPUT));
+    this.interpolateMsg(
+        Blockly.Msg.LANG_TEXT_CONTAINS_INPUT,
+        ['OP', dropdown],
+        ['TEXT', getType('text', utils.INPUT), Blockly.ALIGN_RIGHT],
+        ['PIECE_TEXT', text],
+        ['PIECE', getType('text', utils.INPUT), Blockly.ALIGN_RIGHT],
         Blockly.ALIGN_RIGHT);
-    this.setTooltip(Blockly.Msg.LANG_TEXT_CONTAINS_TOOLTIP);
     this.setInputsInline(false);
+
+    this.setTooltip(function() {
+      var mode = this.getFieldValue('OP');
+      return Blockly.Blocks.text_contains.TOOLTIPS()[mode];
+    }.bind(this));
   },
-  typeblock: [{translatedName: Blockly.Msg.LANG_TEXT_CONTAINS_INPUT_CONTAINS}]
+
+  // TODO: This can be removed after the blockly update b/c validators are
+  // properly triggered on load from XML.
+  domToMutation: function (xmlElement) {
+    var mode = xmlElement.getAttribute('mode');
+    Blockly.Blocks.text_contains.adjustToMode.call(this, mode);
+  },
+
+  mutationToDom: function () {
+    var container = document.createElement('mutation');
+    var savedMode = this.getFieldValue('OP');
+    container.setAttribute('mode', savedMode);
+    return container;
+  },
+
+  typeblock: [
+    {
+      translatedName: Blockly.Msg.LANG_TEXT_CONTAINS_OPERATOR_CONTAINS,
+      dropDown: {
+        titleName: 'OP',
+        value: 'CONTAINS'
+      }
+    },
+    {
+      translatedName: Blockly.Msg.LANG_TEXT_CONTAINS_OPERATOR_CONTAINS_ANY,
+      dropDown: {
+        titleName: 'OP',
+        value: 'CONTAINS_ANY'
+      }
+    },
+    {
+      translatedName: Blockly.Msg.LANG_TEXT_CONTAINS_OPERATOR_CONTAINS_ALL,
+      dropDown: {
+        titleName: 'OP',
+        value: 'CONTAINS_ALL'
+      }
+    }
+  ]
+};
+
+/**
+ * Updates the block's PIECE input to reflect the current mode.
+ * @param {string} mode 
+ * @this {!Blockly.BlockSvg}
+ */
+Blockly.Blocks.text_contains.adjustToMode = function (mode) {
+  var utils = Blockly.Blocks.Utilities;
+  var getType = utils.YailTypeToBlocklyType;
+
+  if (mode == 'CONTAINS') {
+    this.getInput('PIECE')
+        .setCheck(getType('text', utils.INPUT));
+    this.setFieldValue(
+        Blockly.Msg.LANG_TEXT_CONTAINS_INPUT_PIECE,
+        'PIECE_TEXT');
+  } else {
+    this.getInput('PIECE')
+      .setCheck(getType('list', utils.INPUT));
+    this.setFieldValue(
+        Blockly.Msg.LANG_TEXT_CONTAINS_INPUT_PIECE_LIST,
+        'PIECE_TEXT');
+  }
+};
+
+// The order here determines the order in the dropdown
+Blockly.Blocks.text_contains.OPERATORS = function() {
+  return [
+    [Blockly.Msg.LANG_TEXT_CONTAINS_OPERATOR_CONTAINS, 'CONTAINS'],
+    [Blockly.Msg.LANG_TEXT_CONTAINS_OPERATOR_CONTAINS_ANY, 'CONTAINS_ANY'],
+    [Blockly.Msg.LANG_TEXT_CONTAINS_OPERATOR_CONTAINS_ALL, 'CONTAINS_ALL'],
+  ]
+};
+
+Blockly.Blocks.text_contains.TOOLTIPS = function() {
+  return {
+    CONTAINS: Blockly.Msg.LANG_TEXT_CONTAINS_TOOLTIP_CONTAINS,
```

Comment:
For this and `HELPURLS` it might be good to use strings as keys. While technically this will work with simple optimizations, it would be nice if we did everything as if we were going to move to advanced optimizations, which would rewrite the non-quoted keys here.